### PR TITLE
Display Preambles

### PR DESCRIPTION
**Pull request by cmc333333**
Preambles are (currently) an idealized tree structure. They don't carry nearly
as much baggage as regulation data, which means we're free to write idealized
versions of the interface we'd be using in regulations. When we can, we'll
backport these designs.

In this case, we're breaking from the regulation UI by:

* Using a single view to represent "partial" and "chrome" views. Partial views
  are used when making an AJAX request or adding "&partial=true" to the url
* Using "sub_context"/"sub_template" rather than attempting to render the
  sub-content via a mock request
* Use an idealized subtree finding function. As a node label encodes its
  location for preambles, we can find subtrees in log(n) time
* Rather than use a dash-separated url arg like "111-22-a-3-ii", use slashes.
  /preamble/2016_111/ is the parent of /preamble/2016_111/p3, which is the
  parent of preamble/2016_111/p3/p0 and so forth

Still have some annoyances though:

* The presence of HTMLBuilder is like a magic incantation here. It's not at
  all clear what it does or how to call it
* There's an inline style for indenting the TOC. I figure we'll remove when we
  know better what it should look like.

Looks like:
screen shot 2016-03-11 at 12 13 25 am
screen shot 2016-03-11 at 12 13 41 am
screen shot 2016-03-11 at 12 13 52 am
screen shot 2016-03-11 at 12 14 03 am

For eregs/notice-and-comment#14

### CODE REVIEW

File: regulations/views/preamble.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            'markup_page_type': 'reg-section'}
+
+
+class PreambleView(View):
+    """Displays either a notice preamble (or a subtree of that preamble). If
+    using AJAX or specifically requesting, generate only the preamble markup;
+    otherwise wrap it in the appropriate "chrome" """
+    def get(self, request, *args, **kwargs):
+        label_parts = kwargs.get('paragraphs', '').split('/')
+        preamble = ApiReader().preamble(label_parts[0])
```
Comment (jmcarp) [inline regulations/views/preamble.py]: If I'm reading this correctly, `preamble` will be `None` if the request document number isn't found, in which case `generate_html_tree` throws an uncaught exception, and the view blows up. Should this throw a 404 if `preamble` isn't found?

Comment (cmc333333) [inline regulations/views/preamble.py]: Oof, yes. Handling 404s slipped off my radar on this one.

File: regulations/views/preamble.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+def find_subtree(root, label_parts):
+    """Given a nested tree and a label to look for, find the associated node
+    in the tree. Note that, unlike regulations, preamble trees _always_ encode
+    their exact position in the tree."""
+    cursor = root
+    while cursor and cursor['label'] != label_parts:
+        next_cursor = None
+        for child in cursor['children']:
+            if child['label'] == label_parts[:len(child['label'])]:
+                next_cursor = child
```
Comment (jmcarp) [inline regulations/views/preamble.py]: I'm expecting that labels are unique across siblings, but for clarity, do we want to `break` once we assign `next_cursor`? Also, do we need error handling here? We don't seem to check that we actually retrieved the requested label parts at the end, if I'm reading correctly.

**Comment by jmcarp** [comment]
Left another minor comment, but this LGTM--especially the part where we don't have a fake a request to a partial view!

File: regulations/views/preamble.py (hunk 2 above)
Comment (cmc333333) [inline regulations/views/preamble.py]: You're right that labels should be unique -- we'd actually get a consistency error when writing the preamble if this were not the case. We could `break` early, but meh. Unless it's confusing, I don't think the execute cycles (to finish off the child list) is worth the inherently evil `break`.

Correct, if we don't find the label, this implicitly returns None. Seems like another area for a 404.

**Pull request merged**